src/integer_mod_q/mat_zq/inverse.rs

```diff
@@ -0,0 +1,194 @@
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `inverse` function.
+
+use super::MatZq;
+use crate::{
+    integer::{MatZ, Z},
+    traits::{Concatenate, Gcd, GetNumRows},
+};
+use flint_sys::fmpz_mod_mat::fmpz_mod_mat_rref;
+
+impl MatZq {
+    /// Returns the inverse of the matrix if it exists (is square and
+    /// has a determinant unequal to zero) and `None` otherwise.
+    ///
+    /// Note that the modulus is assumed to be prime.
+    /// If it is not, it can happen that the function panics.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let mut matrix = MatZq::from_str("[[1,2],[3,4]] mod 7").unwrap();
+    /// let matrix_invert = matrix.inverse_prime().unwrap();
+    ///
+    /// let id = &matrix_invert * &matrix;
+    ///
+    /// assert_eq!("[[5, 1],[5, 3]] mod 7", matrix_invert.to_string());
+    /// assert_eq!("[[1, 0],[0, 1]] mod 7", id.to_string());
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if the number of rows is not equal to the number of columns.
+    /// - if the inverse of an entry can not be computed because the modulus is not prime.
+    pub fn inverse_prime(&self) -> Option<MatZq> {
+        // Check if determinant is coprime to modulus.
+        let det = MatZ::from(self).det().ok()?;
+        if det.gcd(self.get_mod()) != Z::ONE {
+            None
+        } else {
+            let dimensions = self.get_num_rows();
+
+            // Concatenate the matrix with the identity matrix.
+            let matrix_identity = self
+                .concat_horizontal(&MatZq::identity(dimensions, dimensions, self.get_mod()))
+                .unwrap();
+
+            let identity_inverse = matrix_identity.gaussian_elimination_prime();
+
+            // The inverse is now the right half of the matrix `identity_inverse`.
+            let mut inverse = MatZq::new(dimensions, dimensions, self.get_mod());
+            for i in 0..dimensions {
+                inverse
+                    .set_column(i, &identity_inverse, dimensions + i)
+                    .unwrap();
+            }
+            Some(inverse)
+        }
+    }
+
+    /// Returns the `row echelon form` of the matrix using gaussian elimination or
+    /// panics if the modulus is not prime.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let mut matrix = MatZq::from_str("[[1,2],[3,4]] mod 7").unwrap();
+    /// let matrix_gauss = matrix.gaussian_elimination_prime();
+    ///
+    /// assert_eq!("[[1, 0],[0, 1]] mod 7", matrix_gauss.to_string());
+    /// ```
+    ///
+    /// # Panics ...
+    /// - if the modulus is not prime.
+    pub fn gaussian_elimination_prime(self) -> MatZq {
```

Comment:
Suggestion:
Return a result instead.  It is more rusty and the simplified syntax is not necessary here, since gaussian_elemination_prime will probably not be written in code that often. It also allows programming patterns where the parameter is first thrown into this function, and if an error occurs into the next.

Reply:
For consistency I would say that it should stay as a panic

Reply:
ok...